### ISSUE TITLE
Negative Values after substracitng mean in a

### ISSUE THREAD

**Issue by JonathanPrexl** [Lornatang/ESRGAN-PyTorch#8]
Fist of all: Thank you very much for sharing your code!

I happens from time to time that in the `train_adversarial` function the final tensor in the loss function becomes negative:
`d_loss_hr = adversarial_criterion(hr_output - torch.mean(sr_output), real_label)`
(Mostly when the `torch.mean(sr_output)` is larger then the `hr_output` values)

Is that only a problem on my side, or is that known?

**Comment by Lornatang** [comment]
This is a known issue and correct

**Comment by JonathanPrexl** [comment]
Thanks for your fast answer!

Is there any known workaround or solution for that?

**Comment by qiangwang57** [comment]
> Thanks for your fast answer!
> 
> Is there any known workaround or solution for that?

Hello,

You can check here https://github.com/AlexiaJM/RelativisticGAN, which mentioned if you use BCELoss (which is used in ESRGAN), you need to add sigmoid as the last layer of the Discriminator (which is not included here). However, BCEWithLogitsLoss is recommended as it will do sigmoid implicitly.

**Comment by Vonuo** [comment]
> > Thanks for your fast answer!
> > Is there any known workaround or solution for that?
> 
> Hello,
> 
> You can check here https://github.com/AlexiaJM/RelativisticGAN, which mentioned if you use BCELoss (which is used in ESRGAN), you need to add sigmoid as the last layer of the Discriminator (which is not included here). However, BCEWithLogitsLoss is recommended as it will do sigmoid implicitly.

Hello,
"add sigmoid as the last layer of the Discriminator" can't solve the problem, because "hr_output" will still larger than "sr_output" sometime.

**Comment by Vonuo** [comment]
> This is a known issue and correct

Hello,
Is this problem really corrected? i still get the same problem.

**Comment by Lornatang** [comment]
If you use BCEWithLogits, it will be more stable than SIgmoid+BCELoss,

**Comment by Lornatang** [comment]
This point is not mentioned in the original paper, but it is actually a key factor in the successful training of GAN
